CC-14716: OMS Order commands

kkicheglovspryker opened the pull request:
https://spryker.atlassian.net/browse/CC-14716

kkicheglovspryker committed:
CC-14716 Change unzer-api module version

kkicheglovspryker committed:
CC-14716 Added expenses distributor

kkicheglovspryker committed:
CC-14716 Remove unused method

kkicheglovspryker committed:
Merge branch 'feature/cc-14610/dev-unzer-psp' of github.com:spryker-eco/unzer into feature/cc-14610/cc-14716-order-commands

kkicheglovspryker committed:
CC-14716 Fixed checkout process and expenses distributor

kkicheglovspryker committed:
CC-15652 Fix composer.json

kkicheglovspryker committed:
CC-15652 Fixed CC operations

kkicheglovspryker committed:
CC-15652 Added new repository methods

kkicheglovspryker committed:
CC-14716 Added UnzerBasket v2 and items/shipments distribution

kkicheglovspryker committed:
CC-14716 Added UnzerBasket v2 and items/shipments distribution

kkicheglovspryker committed:
CC-14716 Modified ExpensesDistributor

TomaToSS committed:
feature/CC-15965 fixed missing keypair transfer for basic refund functionality and some data types

TomaToSS committed:
feature/CC-15965 code style fixes

TomaToSS committed:
feature/CC-15965 fixed missing input

kkicheglovspryker committed:
CC-14716 Finished partial charge for regular and marketplace CC

kkicheglovspryker committed:
Merge branch 'feature/cc-14610/cc-14716-order-commands' of github.com:spryker-eco/unzer into feature/cc-14610/cc-14716-order-commands

kkicheglovspryker committed:
CC-14716 Add regular charge request prepare

AsonUnique committed:
CC-14723 Add Sofort and Bank Transfer payment methods.

AsonUnique committed:
CC-14723 Merged feature/cc-14610/cc-14716-order-commands branch

AsonUnique committed:
CC-14723 Adjust const names

AsonUnique committed:
CC-14723 Fix state machine

TomaToSS committed:
feature/CC-15965 added shipment refund functionality
* added config constatnts to set strategy
* added dependecies
* codestyle fixes added

TomaToSS committed:
feature/CC-14339 codestyle fixes added

Commit:
Fixed cr comments

Commit:
Fixed codestyle errors

Commit:
Fixed architecture errors

Commit:
Fixed branch conflicts

michbeck committed:
Merge pull request #12 from spryker-eco/feature/cc-14610/cc-14723-payment-method-sofort-bank-transfer

CC-14723: Unzer payment method Sofort bank transfer

kkicheglovspryker committed:
CC-14716 Minor fixes

Commit:
Merged dev-unzer-psp

Commit:
Fixed duplicated methods

Commit:
Added salutation map and default value

Commit:
Fixed invalid salutation through mapping implementation

Commit:
Fixed setting amounts in UnzerPaymentMapper

Commit:
Renamed UnzerCheckoutMapper to UnzerCheckoutPostSaveMapper

Commit:
Fixed performCreateMarketplaceBasketApiCall method

Commit:
Added constants for payment method installment secured & fixed bank transfer

Commit:
Merged feature/cc-14610/dev-unzer-psp

Commit:
Fixed factories

Commit:
Fixed payment method import

Commit:
Added missing payment method templates

Commit:
Removed unnecessary payment method definition and fixed constraint for paymentResource.id

Commit:
Added subforms for standard / marketplace bank transfer and sofort

Commit:
Removed OrFail for participant id while can be null for standard

kkicheglovspryker committed:
Merge branch 'feature/cc-14610/dev-unzer-psp' of github.com:spryker-eco/unzer into feature/cc-14610/cc-14716-order-commands

kkicheglovspryker committed:
CC-14716 Fix composer.json

Commit:
Fixed UnzerbusinessFactory

Commit:
Added missing AbstractPaymentProcessor & MarketplaceBankTransferProcessor

Commit:
Fixed AbstractPaymentProcessor

Commit:
Fixed MarketplaceBankTransferProcessor

Commit:
Renamed ChargePaymentProcessor to DirectPaymentProcessor

Commit:
Fixed UnzerBusinessFactory::createPaymentProcessorResolver

Commit:
Fixed UnzerMarketplaceRefundProcessor

Commit:
Introduced UnzerRefundProcessor

kkicheglovspryker committed:
CC-14716 Restore correct code

kkicheglovspryker committed:
CC-14716 Fixed factory

kkicheglovspryker committed:
CC-14716 Restore changes

Commit:
Fixed UnzerOmsMapper

kkicheglovspryker committed:
CC-14716 Restore changes

Commit:
Adjusted payment method import

Commit:
Merge remote-tracking branch 'composer/feature/cc-14610/cc-14716-order-commands' into feature/cc-14610/cc-14716-order-commands

Commit:
Introduced PaymentUnzerTransactionConditions.unzerKeypairIds property

Commit:
Updated setPaymentUnzerTransactionFilters to use keypairIds and market method as deprecated

kkicheglovspryker committed:
CC-14716 Fix mapper

kkicheglovspryker committed:
CC-14716 Fix payemnts

kkicheglovspryker committed:
CC-14716 Refactored refund, added expenses refund strategies

kkicheglovspryker committed:
CC-14716 Added CS fixes

kkicheglovspryker committed:
CC-14716 Added CS fixes

michbeck commented on config/Zed/Oms/UnzerBankTransfer01.xml:
```diff
@@ -0,0 +1,68 @@
+            <event name="close" manual="false" timeout="14 days"/>
+        </events>
+
+        <transitions>
+
+            <transition happy="true">
```
```suggestion
        <transitions>
            <transition happy="true">
```

michbeck commented on config/Zed/Oms/UnzerMarketplaceSofort01.xml:
```diff
@@ -0,0 +1,68 @@
+            <event name="close" manual="false" timeout="14 days"/>
+        </events>
+
+        <transitions>
+
+            <transition happy="true">
```
```suggestion
        <transitions>
            <transition happy="true">
```

michbeck commented on config/Zed/Oms/UnzerSofort01.xml:
```diff
@@ -0,0 +1,68 @@
+            <event name="close" manual="false" timeout="14 days"/>
+        </events>
+
+        <transitions>
+
+            <transition happy="true">
```
```suggestion
        <transitions>
            <transition happy="true">
```

michbeck commented on src/SprykerEco/Zed/Unzer/Business/Credentials/UnzerCredentialsValidator/UnzerCredentialsStoreRelationsValidator.php:
```diff
@@ -85,7 +85,7 @@ public function validate(UnzerCredentialsTransfer $unzerCredentialsTransfer): Un
     protected function isUnzerCredentialsHaveStoreRelations(UnzerCredentialsTransfer $unzerCredentialsTransfer): bool
     {
-        if (in_array($unzerCredentialsTransfer->getTypeOrFail(), UnzerConstants::UNZER_CHILD_CONFIG_TYPES, true)) {
+        if (in_array((int)$unzerCredentialsTransfer->getTypeOrFail(), UnzerConstants::UNZER_CHILD_CONFIG_TYPES, true)) {
```
Imho we don't need cast here because getTypeOrFail already returns int or not?

kkicheglovspryker replied:
We need to cast here cause Symfony form in another module builds object in such way so here could be string

michbeck commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerRefundProcessorInterface.php:
```diff
@@ -15,7 +15,7 @@ interface UnzerRefundProcessorInterface
      * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
      * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
-     * @param array $salesOrderItemIds
+     * @param array<int> $salesOrderItemIds
```
```suggestion
     * @param array<int, int> $salesOrderItemIds
```

kkicheglovspryker replied:
I see here plain array w/o indexes

michbeck commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,315 @@
+    /**
+     * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
+     * @param \Generated\Shared\Transfer\PaymentUnzerTransfer $paymentUnzerTransfer
+     * @param \ArrayObject|array<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```
```suggestion
     * @param \ArrayObject<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```

michbeck commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,315 @@
+    /**
+     * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
+     * @param \Generated\Shared\Transfer\PaymentUnzerTransfer $paymentUnzerTransfer
+     * @param \ArrayObject|array<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```
```suggestion
     * @param \ArrayObject<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```

michbeck commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,315 @@
+    }
+
+    /**
+     * @param \ArrayObject|array<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```
```suggestion
     * @param \ArrayObject<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```

michbeck commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,315 @@
+                    ->setAmountGross($expenseTransfer->getRefundableAmountOrFail() / UnzerConstants::INT_TO_FLOAT_DIVIDER),
+            );
+
+        return $unzerRefundTransfer;
```
```suggestion
        return (new UnzerRefundTransfer())
            ->setPaymentId($paymentUnzerTransfer->getPaymentIdOrFail())
            ->setChargeId($expenseTransfer->getUnzerChargeIdOrFail())
            ->addItem(
                (new UnzerRefundItemTransfer())
                    ->setBasketItemReferenceId(
                        sprintf(UnzerConstants::UNZER_MARKETPLACE_BASKET_SHIPMENT_REFERENCE_ID, $expenseTransfer->getUnzerParticipantIdOrFail()),
                    )
                    ->setQuantity(UnzerConstants::PARTIAL_REFUND_QUANTITY)
                    ->setAmountGross($expenseTransfer->getRefundableAmountOrFail() / UnzerConstants::INT_TO_FLOAT_DIVIDER),
            );
```

michbeck commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,315 @@
+
+    /**
+     * @param \Generated\Shared\Transfer\PaymentUnzerTransfer $paymentUnzerTransfer
+     * @param \ArrayObject|array<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```
```suggestion
     * @param \ArrayObject<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```

michbeck commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpanderInterface.php:
```diff
@@ -0,0 +1,28 @@
+    /**
+     * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
+     * @param \Generated\Shared\Transfer\PaymentUnzerTransfer $paymentUnzerTransfer
+     * @param \ArrayObject|array<\Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```
```suggestion
     * @param \ArrayObject<int, \Generated\Shared\Transfer\ExpenseTransfer> $expenseTransfersCollectionForRefund
```

kkicheglovspryker replied:
this collection is not indexed

michbeck commented on src/SprykerEco/Zed/Unzer/Persistence/UnzerRepository.php:
```diff
@@ -113,6 +115,8 @@ public function findPaymentUnzerOrderItemByIdSalesOrderItem(int $idSalesOrderIte
     /**
+     * @deprecated Please remove usage and use UnzerRepository::findPaymentUnzerTransactionCollectionByCriteria
+     *
      * @param string $paymentId
```
If marketplace is not using this anymore we can imho remove it from reader and repository

kkicheglovspryker committed:
CC-14716 Remove deprecated method

romanhavrylko commented on src/SprykerEco/Shared/Unzer/Transfer/unzer.transfer.xml:
```diff
@@ -117,6 +122,15 @@
     </transfer>
 
+    <transfer name="UnzerCharge">
+        <property name="paymentId" type="string"/>
```
```suggestion
        <property name="idPayment" type="string"/>
```

kkicheglovspryker replied:
It is not database-related id. It comes from Unzer API

romanhavrylko commented on src/SprykerEco/Shared/Unzer/Transfer/unzer.transfer.xml:
```diff
@@ -117,6 +122,15 @@
 
+    <transfer name="UnzerCharge">
+        <property name="paymentId" type="string"/>
+        <property name="authorizeId" type="string"/>
```
```suggestion
        <property name="idAuthorize" type="string"/>
```

romanhavrylko commented on src/SprykerEco/Shared/Unzer/Transfer/unzer.transfer.xml:
```diff
@@ -117,6 +122,15 @@
+        <property name="paymentId" type="string"/>
+        <property name="authorizeId" type="string"/>
+        <property name="amount" type="int"/>
+        <property name="orderId" type="string"/>
```
```suggestion
        <property name="idOrder" type="string"/>
```

kkicheglovspryker replied:
It is not database-related id. It comes from Unzer API

romanhavrylko commented on src/SprykerEco/Shared/Unzer/Transfer/unzer.transfer.xml:
```diff
@@ -117,6 +122,15 @@
+        <property name="authorizeId" type="string"/>
+        <property name="amount" type="int"/>
+        <property name="orderId" type="string"/>
+        <property name="invoiceId" type="string"/>
```
```suggestion
        <property name="idInvoice" type="string"/>
```

kkicheglovspryker replied:
It is not database-related id. It comes from Unzer API

romanhavrylko commented on src/SprykerEco/Shared/Unzer/Transfer/unzer.transfer.xml:
```diff
@@ -205,6 +222,13 @@
         <property name="idSalesOrderItem" type="int"/>
         <property name="shipment" type="Shipment"/>
+        <property name="calculatedExpensesCost" type="int"/>
+        <property name="expenseId" type="string"/>
```
```suggestion
        <property name="idExpense" type="string"/>
```

kkicheglovspryker replied:
It is not database-related id. It comes from Unzer API

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,118 @@
+    {
+        $expensesGroupedByParticipantId = [];
+        foreach ($expenses as $expenseTransfer) {
+            $expensesGroupedByParticipantId[$expenseTransfer->getUnzerParticipantId()][] = $expenseTransfer;
```
```suggestion
            $expensesGroupedByParticipantId[$expenseTransfer->getUnzerParticipantIdOrFail()][] = $expenseTransfer;
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,118 @@
+    protected function addGroupedExpensesByParticipantId(UnzerBasketTransfer $unzerBasketTransfer, ArrayObject $expenses): UnzerBasketTransfer
+    {
+        $expensesGroupedByParticipantId = [];
+        foreach ($expenses as $expenseTransfer) {
```
Would it make sense to extract this loop to a separate method for better readability?

AsonUnique replied:
+1
https://spryker.atlassian.net/wiki/spaces/RFC/pages/2099413078/Accepted+RFC+Naming+of+methods+convention+update.+Getting+indexed+grouped+collections.

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,118 @@
+                ->setBasketItemReferenceId($referenceId);
+
+            /** @phpstan-var array<string, \Generated\Shared\Transfer\ExpenseTransfer> $expensesCollection */
+            foreach ($expensesCollection as $expenseTransfer) {
```
Would it make sense to extract this loop to a separate method for better readability?

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Mapper/UnzerPaymentMapper.php:
```diff
@@ -63,10 +65,10 @@ public function mapUnzerPaymentTransferToPaymentUnzerTransfer(
-            ->setAmountRemaining((int)$unzerPaymentTransfer->getAmountRemaining() * UnzerConstants::INT_TO_FLOAT_DIVIDER)
-            ->setAmountCanceled((int)$unzerPaymentTransfer->getAmountCanceled() * UnzerConstants::INT_TO_FLOAT_DIVIDER)
-            ->setAmountCharged((int)$unzerPaymentTransfer->getAmountCharged() * UnzerConstants::INT_TO_FLOAT_DIVIDER)
+            ->setAmountTotal((int)($unzerPaymentTransfer->getAmountTotal() * UnzerConstants::INT_TO_FLOAT_DIVIDER))
```
Do we need type casting here?

kkicheglovspryker replied:
yes, cause of phpstan

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,238 @@
+    protected function createUnzerCharge(
+        UnzerPaymentTransfer $unzerPaymentTransfer,
+        ItemCollectionTransfer $itemCollectionTransfer,
+        ?string $authorizeId = null
```
```suggestion
        ?string $idAuthorize = null
```

kkicheglovspryker replied:
it is Unzer API related name

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,238 @@
+
+        if ($this->countItemsCharged($paymentUnzerOrderItemCollectionTransfer) > 0) {
+            return $unzerChargeTransfer;
+        }
```
```suggestion
        if (
            $orderTransfer->getExpenses()->count() === 0
            || $this->countItemsCharged($paymentUnzerOrderItemCollectionTransfer)
        ) {
            return $unzerChargeTransfer;
        }
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,180 @@
-        $this->unzerReader = $unzerReader;
-        $this->unzerPaymentMapper = $unzerPaymentMapper;
-        $this->unzerChargeAdapter = $unzerChargeAdapter;
+    protected function groupOrderItemsByParticipantId(
```
```suggestion
    protected function getOrderItemsGroupedByIdParticipant(
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,180 @@
+            }
+
+            foreach ($orderTransfer->getItems() as $itemTransfer) {
+                if ($itemTransfer->getIdSalesOrderItem() === $paymentUnzerOrderItem->getIdSalesOrderItem()) {
+                    $itemCollectionTransfer = $orderItemsGroupedByParticipant[$paymentUnzerOrderItem->getParticipantId()] ?? new ItemCollectionTransfer();
+                    $itemCollectionTransfer->addItem($itemTransfer);
+                    $orderItemsGroupedByParticipant[$paymentUnzerOrderItem->getParticipantId()] = $itemCollectionTransfer;
+                    $itemTransfer->setUnzerParticipantId($paymentUnzerOrderItem->getParticipantId());
+
+                    break;
+                }
```
```suggestion
                if ($itemTransfer->getIdSalesOrderItem() !== $paymentUnzerOrderItem->getIdSalesOrderItem()) {
                    continue;
                }
                
                $itemCollectionTransfer = $orderItemsGroupedByParticipant[$paymentUnzerOrderItem->getParticipantId()] ?? new ItemCollectionTransfer();
                 $itemCollectionTransfer->addItem($itemTransfer);
                 $orderItemsGroupedByParticipant[$paymentUnzerOrderItem->getParticipantId()] = $itemCollectionTransfer;
                 $itemTransfer->setUnzerParticipantId($paymentUnzerOrderItem->getParticipantId());

                  break;
                }
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,238 @@
+     *
+     * @return int
+     */
+    protected function countItemsCharged(
```
```suggestion
    protected function getChargedItemsCount(
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,180 @@
+     * @return array<string, string>
      */
-    protected $unzerPaymentMapper;
+    protected function groupAuthorizeIdsByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer): array
```
The result looks more like `indexed`
```suggestion
    protected function getAuthorizeIdsIndexedByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer): array
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,180 @@
+     *
+     * @return int
+     */
+    protected function countItemsChargedByParticipantId(
```
```suggestion
    protected function getChargedItemsCountByIdParticipant(
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,343 @@
+     *
+     * @return array<string, \Generated\Shared\Transfer\ItemCollectionTransfer>
+     */
+    protected function groupRefundItemTransfersByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer, RefundTransfer $refundTransfer): array
```
```suggestion
    protected function getRefundItemsGroupedByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer, RefundTransfer $refundTransfer): array
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,343 @@
+     *
+     * @return \Generated\Shared\Transfer\UnzerRefundTransfer
+     */
+    protected function createUnzerRefund(
```
```suggestion
    protected function createUnzerRefundTransfer(
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,343 @@
+     *
+     * @return string
+     */
+    protected function getChargeIdByParticipantId(
```
```suggestion
    protected function getIdChargeByIdParticipant(
```

kkicheglovspryker replied:
Names came from Unzer API specs

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,343 @@
+        string $participantId
+    ): string {
+        foreach ($paymentUnzerTransactionCollectionTransfer->getPaymentUnzerTransactions() as $paymentUnzerTransactionTransfer) {
+            if ($paymentUnzerTransactionTransfer->getParticipantId() === null) {
+                return $paymentUnzerTransactionTransfer->getTransactionIdOrFail();
+            }
+
+            if ($paymentUnzerTransactionTransfer->getParticipantIdOrFail() === $participantId) {
+                return $paymentUnzerTransactionTransfer->getTransactionIdOrFail();
+            }
```
```suggestion
        if (
            $paymentUnzerTransactionTransfer->getParticipantId() === null
            || $paymentUnzerTransactionTransfer->getParticipantIdOrFail() === $participantId
        ) {
            return $paymentUnzerTransactionTransfer->getTransactionIdOrFail();
        }
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,238 @@
+    {
+        $paymentUnzerTransfer = $this->unzerRepository->findPaymentUnzerByOrderReference($orderTransfer->getOrderReference());
+        if ($paymentUnzerTransfer === null) {
+            throw new UnzerException('Unzer Payment not found for order reference: ' . $orderTransfer->getOrderReference());
```
```suggestion
            throw new UnzerException(sprintf(
               'Unzer Payment not found for order reference: %s',
               $orderTransfer->getOrderReference(),
           ));
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,238 @@
+            ->resolveUnzerCredentialsByCriteriaTransfer($unzerCredentialsCriteriaTransfer);
+
+        if ($unzerCredentialsTransfer === null) {
+            throw new UnzerException('UnzerCredentials not found by keypairId: ' . $keypairId);
```
```suggestion
            throw new UnzerException('UnzerCredentials not found by keypairId: ' . $keypairId);
            throw new UnzerException(sprintf(
                'UnzerCredentials not found by keypairId: %s',
                $keypairId,
            ));
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,180 @@
+    {
+        $paymentUnzerTransfer = $this->unzerRepository->findPaymentUnzerByOrderReference($orderTransfer->getOrderReference());
+        if ($paymentUnzerTransfer === null) {
+            throw new UnzerException('Unzer Payment not found for order reference: ' . $orderTransfer->getOrderReference());
```
```suggestion
            throw new UnzerException(sprintf(
                'Unzer Payment not found for order reference: %s',
                $orderTransfer->getOrderReference(),
            ));
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerRefundProcessor.php:
```diff
@@ -0,0 +1,263 @@
+     *
+     * @return \Generated\Shared\Transfer\UnzerRefundTransfer
+     */
+    protected function createUnzerRefund(
```
```suggestion
    protected function createUnzerRefundTransfer(
```

romanhavrylko commented on src/SprykerEco/Zed/Unzer/UnzerConfig.php:
```diff
@@ -95,19 +95,23 @@ class UnzerConfig extends AbstractBundleConfig
         UnzerConstants::UNZER_PAYMENT_METHOD_SOFORT => [
-            UnzerSharedConfig::PAYMENT_METHOD_KEY_SOFORT,
-            UnzerSharedConfig::PAYMENT_METHOD_KEY_MARKETPLACE_SOFORT,
+             UnzerSharedConfig::PAYMENT_METHOD_KEY_SOFORT,
```
Please, revert

kkicheglovspryker committed:
CC-14716 dev CR fixes

kkicheglovspryker committed:
CC-14716 fix composer scripts

AsonUnique commented on src/SprykerEco/Shared/Unzer/UnzerConfig.php:
```diff
@@ -245,4 +255,19 @@ class UnzerConfig extends AbstractSharedConfig
+    /**
+     * @var array<string, string>
+     */
+    public const SALUTATION_MAP = [
```
Not sure that this map belongs here, maybe it's better to move it to Zed layer config?

AsonUnique commented on src/SprykerEco/Shared/Unzer/UnzerConfig.php:
```diff
@@ -245,4 +255,19 @@ class UnzerConfig extends AbstractSharedConfig
+    /**
+     * @var string
+     */
+    public const SALUTATION_DEFAULT = 'unknown';
```
The same goes to this const

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,130 @@
+        foreach ($expenses as $expenseTransfer) {
+            $unzerBasketItemTransfer->setAmountPerUnit(
+                $unzerBasketItemTransfer->getAmountPerUnit() +
+                $expenseTransfer->getSumGrossPrice() / UnzerConstants::INT_TO_FLOAT_DIVIDER,
```
It will make sense to extract this to a separate method since you have the same calculation in `addExpensesToUnzerBasketItem` method.

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,130 @@
+     */
+    protected function addExpensesToUnzerBasketItem(array $expensesCollection, UnzerBasketItemTransfer $unzerBasketItemTransfer): UnzerBasketItemTransfer
+    {
+        /** @phpstan-var array<string, \Generated\Shared\Transfer\ExpenseTransfer> $expensesCollection */
```
Can you specify it in docblock? maybe you won't need this additional comment here.

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,237 @@
+        }
+
+        $unzerChargeTransfer = $this->createUnzerCharge($unzerPaymentTransfer, $itemCollectionTransfer);
+        $this->addExpensesToUnzerChargeTransfer($unzerChargeTransfer, $orderTransfer, $paymentUnzerOrderItemCollectionTransfer);
```
```suggestion
        $unzerChargeTransfer = $this->addExpensesToUnzerChargeTransfer($unzerChargeTransfer, $orderTransfer, $paymentUnzerOrderItemCollectionTransfer);
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,237 @@
+            foreach ($orderTransfer->getItems() as $itemTransfer) {
+                $itemTransferFkSalesExpense = $itemTransfer->getShipmentOrFail()->getMethodOrFail()->getFkSalesExpense();
+                $expenseTransferFkSalesExpense = $expenseTransfer->getShipmentOrFail()->getMethodOrFail()->getFkSalesExpense();
+                if ($itemTransferFkSalesExpense === $expenseTransferFkSalesExpense) {
```
We usually trying to avoid triple nesting

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,182 @@
+     * @return array<string, string>
      */
-    protected $unzerPaymentMapper;
+    protected function groupAuthorizeIdsIndexedByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer): array
```
```suggestion
    protected function getAuthorizeIdsIndexedByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer): array
```

https://spryker.atlassian.net/wiki/spaces/RFC/pages/2099413078/Accepted+RFC+Naming+of+methods+convention+update.+Getting+indexed+grouped+collections.

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,182 @@
+    {
+        $paymentUnzerTransactionCollectionTransfer = $this->getPaymentUnzerTransactionCollection($paymentUnzerTransfer);
+
+        $authIdsGroupedByParticipant = [];
```
```suggestion
        $indexedAuthorizeTransactionIds = [];
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,182 @@
+        OrderTransfer $orderTransfer,
+        array $salesOrderItemIds
+    ): array {
+        $orderItemsGroupedByParticipant = [];
```
```suggestion
        $groupedOrderItems = [];
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,182 @@
+            foreach ($orderTransfer->getItems() as $itemTransfer) {
+                $itemTransferFkSalesExpense = $itemTransfer->getShipmentOrFail()->getMethodOrFail()->getFkSalesExpense();
+                $expenseTransferFkSalesExpense = $expenseTransfer->getShipmentOrFail()->getMethodOrFail()->getFkSalesExpense();
+                if ($itemTransfer->getUnzerParticipantId() === $participantId && $itemTransferFkSalesExpense === $expenseTransferFkSalesExpense) {
```
Same for triple nesting here

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,182 @@
+     *
+     * @return int
+     */
+    protected function getChargedItemsCountByIdParticipant(
```
Seems like you don't really need the actual count of charged Items, you just need at least 1 charged item.

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/Mapper/UnzerMarketplaceRefundMapper.php:
```diff
@@ -0,0 +1,73 @@
+        return $unzerRefundItemTransfer
+            ->setAmountGross($itemTransfer->getSumPriceToPayAggregationOrFail() * UnzerConstants::INT_TO_FLOAT_DIVIDER)
+            ->setParticipantId($itemTransfer->getUnzerParticipantIdOrFail())
+            ->setAmountGross($itemTransfer->getSumPriceToPayAggregationOrFail() / UnzerConstants::INT_TO_FLOAT_DIVIDER)
```
You call `setAmountGross` twice with different values.

kkicheglovspryker replied:
fixed

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,342 @@
+            $paymentUnzerTransfer->getOrderIdOrFail(),
+        );
+
+        $result = [];
```
```suggestion
        $groupedRefundItems = [];
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,342 @@
+     */
+    protected function findParticipantIdByItem(PaymentUnzerOrderItemCollectionTransfer $paymentUnzerOrderItemsCollection, ItemTransfer $itemTransfer): string
+    {
+        foreach ($paymentUnzerOrderItemsCollection->getPaymentUnzerOrderItems() as $paymentUnzerOrderItem) {
```
What if you won't find participant ID by item? 
I think, method should return `null` in this case

kkicheglovspryker replied:
fixed

AsonUnique replied:
In this case method should be named `getParticipantIdByItem`

kkicheglovspryker replied:
done

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,342 @@
+
+    /**
+     * @param \Generated\Shared\Transfer\PaymentUnzerTransfer $paymentUnzerTransfer
+     * @param array|string $participantIds
```
```suggestion
     * @param array<string> $participantIds
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Quote/UnzerParticipantIdQuoteExpander.php:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @return \Generated\Shared\Transfer\ItemTransfer
+     */
+    protected function setQuoteItemParticipantId(ItemTransfer $itemTransfer, UnzerCredentialsTransfer $unzerCredentialsTransfer): ItemTransfer
```
```suggestion
    protected function addPartisipantIdToItemTransfer(ItemTransfer $itemTransfer, UnzerCredentialsTransfer $unzerCredentialsTransfer): ItemTransfer
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Quote/UnzerParticipantIdQuoteExpander.php:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @return \Generated\Shared\Transfer\ExpenseTransfer
+     */
+    protected function setQuoteExpenseParticipantId(ExpenseTransfer $expenseTransfer, UnzerCredentialsTransfer $unzerCredentialsTransfer): ExpenseTransfer
```
```suggestion
    protected function addPartisipantIdToExpenseTransfer(ExpenseTransfer $expenseTransfer, UnzerCredentialsTransfer $unzerCredentialsTransfer): ExpenseTransfer
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Quote/UnzerParticipantIdQuoteExpander.php:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @return string|null
+     */
+    protected function getMerchantParticipantId(int $parentIdUnzerCredentials, string $merchantReference): ?string
```
This can cause problems in case of a big number of items/expenses, maybe we should consider creating some kind of cache for participant ids?

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Quote/UnzerParticipantIdQuoteExpander.php:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @return string|null
+     */
+    protected function getMainMerchantParticipantId(int $parentIdUnzerCredentials): ?string
```
same here

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastOrderItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,104 @@
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param array<int> $salesOrderItemIds
+     *
+     * @return \ArrayObject
```
```suggestion
     * @return \ArrayObject<\Generated\Shared\Transfer\ExpenseTransfer>
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @return \Generated\Shared\Transfer\RefundTransfer
+     */
+    public function prepareUnzerRefund(RefundTransfer $refundTransfer, OrderTransfer $orderTransfer, array $salesOrderItemIds): RefundTransfer
```
```suggestion
    public function prepareUnzerRefundTransfer(RefundTransfer $refundTransfer, OrderTransfer $orderTransfer, array $salesOrderItemIds): RefundTransfer
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,145 @@
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param array<int> $salesOrderItemIds
+     *
+     * @return \ArrayObject
```
```suggestion
     * @return \ArrayObject<\Generated\Shared\Transfer\ExpenseTransfer>
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,145 @@
+        $paymentUnzerOrderItemCollection = $this->unzerRepository->getPaymentUnzerOrderItemCollectionByOrderId($orderTransfer->getOrderReferenceOrFail());
+        $idsSalesShipmentForRefund = $this->detectShipmentsForRefund($orderItemsGroupedByIdSalesShipment, $paymentUnzerOrderItemCollection, $salesOrderItemIds);
+
+        $result = new ArrayObject();
```
```suggestion
        $expenseTransfers = new ArrayObject();
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,145 @@
+            throw new UnzerException(sprintf('Unzer payment for order reference %s not found.', $orderTransfer->getOrderReference()));
+        }
+
+        $expenseTransfersCollectionForRefund = $this->collectExpenseTransfersForRefund($orderTransfer, $salesOrderItemIds);
```
```suggestion
        $expenseTransfersForRefund = $this->collectExpenseTransfersForRefund($orderTransfer, $salesOrderItemIds);
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @return array<int, array<\Generated\Shared\Transfer\ItemTransfer>>
+     */
+    protected function groupOrderItemsByIdSalesShipment(OrderTransfer $orderTransfer)
```
```suggestion
    protected function getOrderItemsGroupedByIdSalesShipment(OrderTransfer $orderTransfer)
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    /**
+     * @param array $orderItemsGroupedByIdSalesShipment
```
```suggestion
     * @param array<int, array<\Generated\Shared\Transfer\ItemTransfer>> $orderItemsGroupedByIdSalesShipment
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,145 @@
+        foreach ($orderItemsGroupedByIdSalesShipment as $idSalesShipment => $itemTransfers) {
+            $totalItemsCount = count($itemTransfers);
+            foreach ($itemTransfers as $itemTransfer) {
+                /** @var \Generated\Shared\Transfer\ItemTransfer $itemTransfer */
```
This one won't be needed after fixing a dockblock

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * @param array $orderItemsGroupedByIdSalesShipment
+     * @param \Generated\Shared\Transfer\PaymentUnzerOrderItemCollectionTransfer $paymentUnzerOrderItemCollectionTransfer
+     * @param array $salesOrderItemIds
```
```suggestion
     * @param array<int> $salesOrderItemIds
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,313 @@
+        $childMarketplaceUnzerCredentialsCollectionTransfer = $this->getChildUnzerCredentialsCollection($mainMarketplaceUnzerCredentialsTransfer);
+
+        foreach ($expenseTransfersCollectionForRefund as $expenseTransfer) {
+            /** @var \Generated\Shared\Transfer\ExpenseTransfer $expenseTransfer */
```
please, double-check if you need it.

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,313 @@
+    ): UnzerRefundTransfer {
+        $refundAmountTotal = 0;
+        foreach ($expenseTransfersCollectionForRefund as $expenseTransfer) {
+            /** @var \Generated\Shared\Transfer\ExpenseTransfer $expenseTransfer */
```
same here

AsonUnique commented on src/SprykerEco/Zed/Unzer/Persistence/UnzerRepository.php:
```diff
@@ -301,4 +265,43 @@ protected function setUnzerConfigFilters(
+     *
+     * @return \Orm\Zed\Unzer\Persistence\SpyPaymentUnzerTransactionQuery
+     */
+    protected function setPaymentUnzerTransactionFilters(
```
```suggestion
    protected function applyPaymentUnzerTransactionConditions(
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Persistence/UnzerRepository.php:
```diff
@@ -243,6 +185,28 @@ public function findUnzerCredentialsCollectionByCriteria(
+     *
+     * @return \Generated\Shared\Transfer\PaymentUnzerTransactionCollectionTransfer
+     */
+    public function findPaymentUnzerTransactionCollectionByCriteria(
```
```suggestion
    public function getPaymentUnzerTransactionCollectionByCriteria(
```

kkicheglovspryker committed:
CC-14716 Fix dev CR notices

kkicheglovspryker committed:
CC-14716 Fix dev CR notices

kkicheglovspryker committed:
CC-14716 Fix dev CR notices

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @return \Generated\Shared\Transfer\UnzerBasketTransfer
+     */
+    protected function addExpensesGroupedByParticipantId(UnzerBasketTransfer $unzerBasketTransfer, ArrayObject $expenses): UnzerBasketTransfer
```
```suggestion
    protected function addExpensesGroupedByParticipantId(UnzerBasketTransfer $unzerBasketTransfer, ArrayObject $expenseTransfers): UnzerBasketTransfer
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @return array<string, \Generated\Shared\Transfer\ExpenseTransfer>
+     */
+    protected function getExpensesGroupedByParticipantId(ArrayObject $expenses): array
```
```suggestion
    protected function getExpensesGroupedByParticipantId(ArrayObject $expenseTransfers): array
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @return \Generated\Shared\Transfer\UnzerBasketTransfer
+     */
+    protected function addStandardExpenses(UnzerBasketTransfer $unzerBasketTransfer, ArrayObject $expensesCollection): UnzerBasketTransfer
```
```suggestion
    protected function addStandardExpenses(UnzerBasketTransfer $unzerBasketTransfer, ArrayObject $expenseTransfers): UnzerBasketTransfer
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @return \Generated\Shared\Transfer\UnzerBasketItemTransfer
+     */
+    protected function addExpensesToUnzerBasketItem(array $expensesCollection, UnzerBasketItemTransfer $unzerBasketItemTransfer): UnzerBasketItemTransfer
```
```suggestion
    protected function addExpensesToUnzerBasketItem(array $expenseTransfers, UnzerBasketItemTransfer $unzerBasketItemTransfer): UnzerBasketItemTransfer
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastOrderItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,104 @@
+    /**
+     * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param array $salesOrderItemIds
```
```suggestion
     * @param array<int> $salesOrderItemIds
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param array $salesOrderItemIds
```
```suggestion
     * @param array<int> $salesOrderItemIds
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/NoExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,26 @@
+    /**
+     * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param array $salesOrderItemIds
```
```suggestion
     * @param array<int> $salesOrderItemIds
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/UnzerExpensesRefundStrategyInterface.php:
```diff
@@ -0,0 +1,23 @@
+    /**
+     * @param \Generated\Shared\Transfer\RefundTransfer $refundTransfer
+     * @param \Generated\Shared\Transfer\OrderTransfer $orderTransfer
+     * @param array $salesOrderItemIds
```
```suggestion
     * @param array<int> $salesOrderItemIds
```

AsonUnique commented on src/SprykerEco/Zed/Unzer/UnzerConfig.php:
```diff
@@ -14,9 +14,9 @@
-     */
+ /**
+  * @var array<int, string>
+  */
```
Please, fix the intendation.

kkicheglovspryker committed:
CC-14716 Fix dev CR notices

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,346 @@
+            throw new UnzerException(sprintf('Unzer payment for order reference %s not found.', $orderTransfer->getOrderReference()));
+        }
+
+        $groupRefundItemTransfersByParticipantIds = $this->getRefundItemsGroupedByParticipantIds($paymentUnzerTransfer, $refundTransfer);
```
```suggestion
        $groupedRefundItemTransfers = $this->getRefundItemsGroupedByParticipantIds($paymentUnzerTransfer, $refundTransfer);
```

Commit:
Fixed unzer databuilder
Updated codeception configuration
Fixed / Updated tests
Fixed composer.json
Fixed phpstan.neon

Commit:
Fixed unzer transfer errors
Fixed phpstan errors
Renamed MarketplaceBankTransferProcessor to MarketplaceDirectPaymentProcessor

Commit:
Merged origin/cc-14716-order-commands

FRITS-Kh commented on src/SprykerEco/Yves/Unzer/Theme/default/views/sofort/sofort.twig:
```diff
@@ -0,0 +1,9 @@
+{% extends model('template') %}
+
+{% define data = {
+    form: form
+} %}
+
```
Please add a template for all of these views to avoid duplication.

michbeck replied:
As discussed, not changed.

Commit:
Fixed composer.json

Commit:
Fixed transfer errors

Commit:
Fixed composer dependency violations

Commit:
Fixed composer dependency violations

Commit:
Added spryker-shop/shop-ui dependency

Commit:
Fixed phpstan errors
Fixed UnzerBasketItemTransfer vat value

romanhavrylko commented on src/SprykerEco/Shared/Unzer/Transfer/unzer.transfer.xml:
```diff
@@ -150,6 +164,7 @@
     <transfer name="Message">
         <property name="value" type="string"/>
         <property name="message" type="string"/>
+        <property name="parameters" type="array" singular="parameters"/>
```
```suggestion
        <property name="parameters" type="array" singular="parameter"/>
```

michbeck replied:
This definition is already used in many other modules. So we can't change that.

FRITS-Kh reviewed (approved):
FE part is OK

Commit:
Fixed phpstan errors

Commit:
Fixed phpstan errors

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @return \Generated\Shared\Transfer\UnzerRefundTransfer
+     */
+    protected function createMarketplaceUnzerRefund(PaymentUnzerTransfer $paymentUnzerTransfer, ExpenseTransfer $expenseTransfer): UnzerRefundTransfer
```
```suggestion
    protected function createMarketplaceUnzerRefundTransfer(PaymentUnzerTransfer $paymentUnzerTransfer, ExpenseTransfer $expenseTransfer): UnzerRefundTransfer
```

michbeck replied:
Fixed.

romanhavrylko commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * @return \Generated\Shared\Transfer\UnzerRefundTransfer
+     */
+    protected function createStandardUnzerRefund(
```
```suggestion
    protected function createStandardUnzerRefundTransfer(
```

michbeck replied:
Fixed.

romanhavrylko commented on tests/SprykerEcoTest/Zed/Unzer/Business/ExecutePostSaveHookFacadeTest.php:
```diff
@@ -7,55 +7,296 @@
+ */
 class ExecutePostSaveHookFacadeTest extends UnzerFacadeBaseTest
 {
+//    /**
```
Can it be removed or uncommented?

michbeck replied:
Fixed.

romanhavrylko commented on tests/SprykerEcoTest/Zed/Unzer/Business/ExecutePostSaveHookFacadeTest.php:
```diff
@@ -7,55 +7,296 @@
+            ->setPayment($paymentTransfer);
+        $checkoutResponseTransfer = $this->tester->createCheckoutResponseTransfer();
+
+        $saveOrderTransfer = $this->tester->haveOrder(
+            [
+            'unitPrice' => 72350,
+            'sumPrice' => 72350,
+            'orderReference' => 'DE--2',
+            ],
+            'UnzerSofort01',
+        );
```
```suggestion
        $saveOrderTransfer = $this->tester->haveOrder(
            [
                'unitPrice' => 72350,
                'sumPrice' => 72350,
                'orderReference' => 'DE--2',
            ],
            'UnzerSofort01',
        );
```

Please, check similar cases below

michbeck replied:
Fixed.

romanhavrylko commented on tests/SprykerEcoTest/Zed/Unzer/Business/ExecutePostSaveHookFacadeTest.php:
```diff
@@ -7,55 +7,296 @@
+        $this->assertSame(UnzerZedTester::UNZER_REDIRECT_URL, $checkoutResponseTransfer->getRedirectUrl());
+    }
+
+//    /**
```
Can it be removed or uncommented?

michbeck replied:
Fixed.

Commit:
Fixed comments from CR

Commit:
Fixed codestyle errors

Commit:
Removed blank line

DmitryLymarenko commented on config/Zed/Oms/UnzerBankTransfer01.xml:
```diff
@@ -0,0 +1,67 @@
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="spryker:oms-01 http://static.spryker.com/oms-01.xsd">
+
+    <process name="UnzerBankTransfer01" main="true">
```
naming looks strange
do we really need `Transfer` postfix?
same for file name

kkicheglovspryker replied:
BankTransfer - is a type of payment inside Unzer. It it not related to transfer objects.

DmitryLymarenko replied:
I see now

DmitryLymarenko commented on config/Zed/Oms/UnzerMarketplaceBankTransfer01.xml:
```diff
@@ -31,12 +31,12 @@
             </transition>
 
-            <transition condition="Unzer/IsPaymentCompleted"  happy="true">
+            <transition condition="Unzer/IsPaymentCompleted" happy="true">
```
same question to file name

DmitryLymarenko commented on src/SprykerEco/Shared/Unzer/Transfer/unzer.transfer.xml:
```diff
@@ -254,6 +287,21 @@
+    <transfer name="PaymentUnzerTransactionConditions">
+        <property name="ids" type="int[]" singular="id"/>
+        <property name="participantIds" type="string[]" singular="participantId"/>
+        <property name="fkPaymentUnzerIds" type="int[]" singular="fkPaymentUnzerId"/>
```
```suggestion
        <property name="paymentUnzerIds" type="int[]" singular="paymentUnzerId"/>
```

DmitryLymarenko replied:
according to `filterByFkPaymentUnzer_In` usage at code

DmitryLymarenko commented on src/SprykerEco/Shared/Unzer/Transfer/unzer.transfer.xml:
```diff
@@ -627,6 +691,13 @@
+    </transfer>
+
+    <transfer name="ShipmentMethod">
+        <property name="fkSalesExpense" type="int"/>
```
```suggestion
        <property name="idExpense" type="int"/>
```
or even `ShipmentMethod::expense::idExpense`

transfers should not contain persistence-like fields

kkicheglovspryker replied:
It is not a new property, we added its use from parent transfer

DmitryLymarenko commented on src/SprykerEco/Shared/Unzer/UnzerConfig.php:
```diff
@@ -231,6 +236,11 @@ class UnzerConfig extends AbstractSharedConfig
     public const PAYMENT_METHOD_KEY_WECHAT_PAY = self::PAYMENT_PROVIDER_TYPE . 'WeChatPay';
 
+    /**
+     * @var string
```
`@api` tag is missing
same for similar constants at this config class

kkicheglovspryker replied:
fixed

DmitryLymarenko commented on src/SprykerEco/Shared/Unzer/UnzerConstants.php:
```diff
@@ -47,6 +47,11 @@ interface UnzerConstants
     public const VAULT_DATA_TYPE = 'UNZER:VAULT_DATA_TYPE';
 
+    /**
+     * @var string
```
environmant constants must be marked with `@api` tag

see https://docs.spryker.com/docs/scos/dev/architecture/module-api/definition-of-module-api.html

kkicheglovspryker replied:
fixed

DmitryLymarenko commented on src/SprykerEco/Yves/Unzer/Controller/CallbackController.php:
```diff
@@ -68,7 +68,7 @@ public function paymentResultAction(): RedirectResponse
     public function notificationAction(Request $request): JsonResponse
     {
-        $requestData = $this->getFactory()->getUtilEncodingService()->decodeJson($request->getContent(), true);
+        $requestData = $this->getFactory()->getUtilEncodingService()->decodeJson((string)$request->getContent(), true);
```
what is the reason to cast types here?
you will always get string
probably it worth do declare a variable with `@var` block
or use `@phpstan-var`

DmitryLymarenko commented on src/SprykerEco/Yves/Unzer/Form/MarketplaceBankTransferSubForm.php:
```diff
@@ -7,49 +7,37 @@
      * @var string
      */
-    public const PAYMENT_METHOD = 'marketplace-bank-transfer';
+    public const TEMPLATE_VIEW_PATH = 'views/marketplace-bank-transfer/marketplace-bank-transfer';
```
```suggestion
    protected const TEMPLATE_VIEW_PATH = 'views/marketplace-bank-transfer/marketplace-bank-transfer';
```

DmitryLymarenko commented on src/SprykerEco/Yves/Unzer/Form/MarketplaceBankTransferSubForm.php:
```diff
@@ -7,49 +7,37 @@
-        $resolver->setDefaults([
-            'data_class' => UnzerPaymentTransfer::class,
-        ])->setRequired(static::OPTIONS_FIELD_NAME);
+        return UnzerConfig::PAYMENT_PROVIDER_NAME . DIRECTORY_SEPARATOR . static::TEMPLATE_VIEW_PATH;
```
what is the reason to have this method
it's completely the same as for parent class

DmitryLymarenko commented on src/SprykerEco/Yves/Unzer/Form/MarketplaceSofortSubForm.php:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * @return string
+     */
+    protected function getTemplatePath(): string
```
same

DmitryLymarenko commented on src/SprykerEco/Yves/Unzer/Form/SofortSubForm.php:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * @return string
+     */
+    protected function getTemplatePath(): string
```
same
please check all classes extending `AbstractUnzerSubForm`

DmitryLymarenko replied:
please also check AbstractUnzerSubForm
it looks like `static::OPTIONS_FIELD_NAME`  is used but never declared

kkicheglovspryker replied:
It comes from `SubFormInterface`

DmitryLymarenko replied:
didn't run composer install so phpstorm said this contant is not defined

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/ApiAdapter/Mapper/UnzerGetPaymentMethodsMapper.php:
```diff
@@ -38,8 +38,10 @@ public function mapUnzerApiGetPaymentMethodsResponseTransferToPaymentMethodsTran
-            $paymentMethodsTransfer = $this->mapPaymentMethodKeysToPaymentMethodTransfers($paymentMethodsTransfer, $paymentMethodKeys);
+            $paymentMethodsTransfer = $this->mapPaymentMethodKeysToPaymentMethodTransfers(
+                $paymentMethodsTransfer,
+                $this->unzerConfig->getPaymentMethodKeys($unzerApiPaymentMethodTransfer->getPaymentMethodKeyOrFail()),
```
a bit unclear to me what happens here
getting payment method keys by payment method key
probably naming could be better

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+    /**
+     * @var int
+     */
+    protected const DEFAULT_ZERO_AMOUNT = 0;
```
do we really need `ZERO` at constant name?

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+     */
+    public function distributeExpensesBetweenQuoteItems(QuoteTransfer $quoteTransfer, UnzerBasketTransfer $unzerBasketTransfer): UnzerBasketTransfer
+    {
+        $expensesCollection = $quoteTransfer->getExpenses();
```
```suggestion
        $expensesTransfers = $quoteTransfer->getExpenses();
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+     */
+    protected function addExpensesGroupedByParticipantId(UnzerBasketTransfer $unzerBasketTransfer, ArrayObject $expenseTransfers): UnzerBasketTransfer
+    {
+        $expensesGroupedByParticipantId = $this->getExpensesGroupedByParticipantId($expenseTransfers);
```
please extract this call to method above
already grouped transfers could be provided as argument instead

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+    {
+        $expensesGroupedByParticipantId = $this->getExpensesGroupedByParticipantId($expenseTransfers);
+
+        foreach ($expensesGroupedByParticipantId as $participantId => $expensesCollection) {
```
```suggestion
        foreach ($expensesGroupedByParticipantId as $participantId => $expenseTransfers) {
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @return array<string, array<int, \Generated\Shared\Transfer\ExpenseTransfer>>
+     */
+    protected function getExpensesGroupedByParticipantId(ArrayObject $expenseTransfers): array
```
```suggestion
    protected function getExpenseTransfersGroupedByParticipantId(ArrayObject $expenseTransfers): array
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Checkout/ExpensesDistributor/UnzerExpensesDistributor.php:
```diff
@@ -0,0 +1,135 @@
+     */
+    protected function getExpensesGroupedByParticipantId(ArrayObject $expenseTransfers): array
+    {
+        $expensesGroupedByParticipantId = [];
```
```suggestion
        $expensesGroupedByParticipantId = [];
```
```suggestion
        $expenseTransfersGroupedByParticipantId = [];
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,245 @@
+    {
+        $paymentUnzerTransfer = $this->unzerRepository->findPaymentUnzerByOrderReference($orderTransfer->getOrderReferenceOrFail());
+        if ($paymentUnzerTransfer === null) {
+            throw new UnzerException(sprintf('Unzer Payment not found for order reference %s', $orderTransfer->getOrderReferenceOrFail()));
```
this should be mentioned at api specs

kkicheglovspryker replied:
fixed

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,245 @@
+
+        $itemCollectionTransfer = new ItemCollectionTransfer();
+        foreach ($orderTransfer->getItems() as $itemTransfer) {
+            if (in_array($itemTransfer->getIdSalesOrderItemOrFail(), $salesOrderItemIds, true)) {
```
please make sure you listed all required transfer props at api specification

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerCreditCardChargeProcessor.php:
```diff
@@ -0,0 +1,245 @@
+            if ($itemTransferFkSalesExpense === $expenseTransferFkSalesExpense) {
+                $expensesAmount += $expenseTransfer->getSumGrossPrice();
+
+                break;
```
`return` could be used instead of `break`

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,200 @@
+    public function charge(OrderTransfer $orderTransfer, array $salesOrderItemIds): void
+    {
+        $paymentUnzerTransfer = $this->unzerRepository->findPaymentUnzerByOrderReference($orderTransfer->getOrderReferenceOrFail());
+        if ($paymentUnzerTransfer === null) {
```
same comments as above

kkicheglovspryker replied:
fixed

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,200 @@
+                $groupedOrderItems[$paymentUnzerOrderItem->getParticipantId()] = $itemCollectionTransfer;
+                $itemTransfer->setUnzerParticipantId($paymentUnzerOrderItem->getParticipantId());
+
+                break;
```
return could be used instead

kkicheglovspryker replied:
No, It will change business logic

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/Mapper/UnzerMarketplaceRefundMapper.php:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @return \Generated\Shared\Transfer\UnzerRefundItemTransfer
+     */
+    public function mapItemTransferToUnzerRefundItemTransfer(
```
looks like protected

DmitryLymarenko replied:
any updates here?

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,346 @@
+        UnzerPaymentMapperInterface $unzerPaymentMapper,
+        UnzerPaymentAdapterInterface $unzerPaymentAdapter,
+        UnzerPaymentSaverInterface $unzerPaymentSaver
+    ) {
```
this class has too much dependenices
we need to refactor it to have not more then 5-6

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,346 @@
+            throw new UnzerException(sprintf('Unzer payment for order reference %s not found.', $orderTransfer->getOrderReferenceOrFail()));
+        }
+
+        $groupRefundItemTransfersByParticipantIds = $this->getRefundItemsGroupedByParticipantIds($paymentUnzerTransfer, $refundTransfer);
```
```suggestion
        $refundItemTransfersGroupedByParticipantId = $this->getRefundItemsGroupedByParticipantIds($paymentUnzerTransfer, $refundTransfer);
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,346 @@
+    {
+        $paymentUnzerTransfer = $this->unzerRepository->findPaymentUnzerByOrderReference($orderTransfer->getOrderReferenceOrFail());
+        if ($paymentUnzerTransfer === null) {
+            throw new UnzerException(sprintf('Unzer payment for order reference %s not found.', $orderTransfer->getOrderReferenceOrFail()));
```
update API specs with exceptions and required transfer fields

kkicheglovspryker replied:
done

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerMarketplaceRefundProcessor.php:
```diff
@@ -0,0 +1,346 @@
+     *
+     * @return array<string, \Generated\Shared\Transfer\ItemCollectionTransfer>
+     */
+    protected function getRefundItemsGroupedByParticipantIds(PaymentUnzerTransfer $paymentUnzerTransfer, RefundTransfer $refundTransfer): array
```
```suggestion
    protected function getRefundItemsGroupedByParticipantId(PaymentUnzerTransfer $paymentUnzerTransfer, RefundTransfer $refundTransfer): array
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerRefundProcessor.php:
```diff
@@ -0,0 +1,263 @@
+        UnzerPaymentAdapterInterface $unzerPaymentAdapter,
+        UnzerPaymentSaverInterface $unzerPaymentSaver
+    ) {
+        $this->unzerCredentialsResolver = $unzerCredentialsResolver;
```
the same comments as for class above

kkicheglovspryker replied:
fixed

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Quote/UnzerParticipantIdQuoteExpander.php:
```diff
@@ -0,0 +1,183 @@
+    /**
+     * @var string
+     */
+    protected $mainMerchantParticipantId;
```
business models must be as much stateless as possible
please avoid these properties usage

kkicheglovspryker replied:
Here we introduced state knowingly. We need to "cache" this one and next property while object exists.
In other way we need to increase method's parameters count.

DmitryLymarenko replied:
discussed at slack

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Quote/UnzerParticipantIdQuoteExpander.php:
```diff
@@ -0,0 +1,183 @@
+     *
+     * @return string|null
+     */
+    protected function getMerchantParticipantIdByMerchantReference(string $merchantReference): ?string
```
```suggestion
    protected function extractMerchantParticipantIdByMerchantReference(string $merchantReference): ?string
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Quote/UnzerParticipantIdQuoteExpander.php:
```diff
@@ -0,0 +1,183 @@
+     */
+    public function expandQuoteItemsWithParticipantIds(QuoteTransfer $quoteTransfer): QuoteTransfer
+    {
+        $marketplaceMainUnzerCredentials = $this->getMarketplaceMainUnzerCredentials($quoteTransfer->getStoreOrFail());
```
```suggestion
        $unzerCredentialsTransfer = $this->getMarketplaceMainUnzerCredentials($quoteTransfer->getStoreOrFail());
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/AbstractExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,36 @@
+use Generated\Shared\Transfer\PaymentUnzerOrderItemCollectionTransfer;
+use SprykerEco\Zed\Unzer\UnzerConstants;
+
+abstract class AbstractExpensesRefundStrategy
```
please avoid inheritence here
this complicates code extending for customers
use composition instead, or just duplicate code
there is no reason for this class to exist as abstract parent

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/ExpensesRefundStrategyResolver/UnzerExpensesRefundStrategyResolver.php:
```diff
@@ -0,0 +1,43 @@
+use SprykerEco\Zed\Unzer\Business\Exception\UnzerException;
+use SprykerEco\Zed\Unzer\Business\Refund\RefundStrategy\UnzerExpensesRefundStrategyInterface;
+
+class UnzerExpensesRefundStrategyResolver implements UnzerExpensesRefundStrategyResolverInterface
```
```suggestion
class UnzerExpenseRefundStrategyResolver implements UnzerExpenseRefundStrategyResolverInterface
```

class names are ususlly singular

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/ExpensesRefundStrategyResolver/UnzerExpensesRefundStrategyResolver.php:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * @var array<int, \Closure>
+     */
+    protected $unzerRefundStrategiesCollection;
```
```suggestion
    protected $unzerRefundStrategies;
```
collection usually means something specific and not regular array

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,144 @@
+        PaymentUnzerOrderItemCollectionTransfer $paymentUnzerOrderItemCollectionTransfer,
+        array $salesOrderItemIds
+    ): array {
+        $result = [];
```
avoid such naming

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,144 @@
+     */
+    protected function getOrderItemsGroupedByIdSalesShipment(OrderTransfer $orderTransfer)
+    {
+        $indexedItems = [];
```
```suggestion
        $groupedItemTransfers = [];
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/LastShipmentItemExpensesRefundStrategy.php:
```diff
@@ -0,0 +1,144 @@
+use SprykerEco\Zed\Unzer\Business\Refund\UnzerRefundExpanderInterface;
+use SprykerEco\Zed\Unzer\Persistence\UnzerRepositoryInterface;
+
+class LastShipmentItemExpensesRefundStrategy extends AbstractExpensesRefundStrategy implements UnzerExpensesRefundStrategyInterface
```
```suggestion
class LastShipmentItemExpenseRefundStrategy extends AbstractExpensesRefundStrategy implements UnzerExpensesRefundStrategyInterface
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/UnzerExpensesRefundStrategyInterface.php:
```diff
@@ -0,0 +1,23 @@
+use Generated\Shared\Transfer\OrderTransfer;
+use Generated\Shared\Transfer\RefundTransfer;
+
+interface UnzerExpensesRefundStrategyInterface
```
```suggestion
interface UnzerExpenseRefundStrategyInterface
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/RefundStrategy/UnzerExpensesRefundStrategyInterface.php:
```diff
@@ -0,0 +1,23 @@
+     *
+     * @return \Generated\Shared\Transfer\RefundTransfer
+     */
+    public function prepareUnzerRefundTransfer(RefundTransfer $refundTransfer, OrderTransfer $orderTransfer, array $salesOrderItemIds): RefundTransfer;
```
this strategy resolving must be mentioned at all affected API specifications

kkicheglovspryker replied:
added

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Refund/UnzerRefundExpander.php:
```diff
@@ -0,0 +1,313 @@
+        $childMarketplaceUnzerCredentialsCollectionTransfer = $this->getChildUnzerCredentialsCollection($mainMarketplaceUnzerCredentialsTransfer);
+
+        foreach ($expenseTransfersCollectionForRefund as $expenseTransfer) {
+            foreach ($childMarketplaceUnzerCredentialsCollectionTransfer->getUnzerCredentials() as $unzerCredentialsTransfer) {
```
inner foreach loop must be extracted to a separate method

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/UnzerBusinessFactory.php:
```diff
@@ -569,15 +667,86 @@ public function createUnzerMarketplaceCreditCardChargeProcessor(): UnzerChargePr
+    public function createUnzerExpensesRefundStrategyResolver(): UnzerExpensesRefundStrategyResolverInterface
+    {
+        return new UnzerExpensesRefundStrategyResolver([
+            UnzerConstants::NO_EXPENSES_REFUND_STRATEGY => function () {
```
please extract this strategies array forming to a separate method

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/UnzerBusinessFactory.php:
```diff
@@ -569,15 +667,86 @@ public function createUnzerMarketplaceCreditCardChargeProcessor(): UnzerChargePr
+    /**
+     * @return \SprykerEco\Zed\Unzer\Business\Payment\Processor\Refund\Mapper\UnzerMarketplaceRefundMapperInterface
+     */
+    public function createUnzerMarketplaceMapper(): UnzerMarketplaceRefundMapperInterface
```
```suggestion
    public function createUnzerMarketplaceRefundMapper(): UnzerMarketplaceRefundMapperInterface
```

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Communication/UnzerCommunicationFactory.php:
```diff
@@ -72,24 +74,32 @@ public function getSalesFacade(): UnzerToSalesFacadeInterface
+     */
+    public function getShipmentFacade(): UnzerToShipmentFacadeInterface
+    {
+        /** @var \SprykerEco\Zed\Unzer\Dependency\UnzerToShipmentFacadeInterface $shipmentFacade */
```
is it really needed?

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Persistence/Mapper/UnzerPersistenceMapper.php:
```diff
@@ -254,4 +255,26 @@ public function mapUnzerCredentialsEntityCollectionToUnzerCredentialsTransferCol
     }
+
+    /**
+     * @param \Propel\Runtime\Collection\ObjectCollection $paymentUnzerTransactionEntities
```
typehint is not precise
please specifiy exact propel entities inside

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Persistence/UnzerRepository.php:
```diff
@@ -301,4 +264,47 @@ protected function setUnzerConfigFilters(
+     */
+    protected function applyPaymentUnzerTransactionConditions(
+        SpyPaymentUnzerTransactionQuery $paymentUnzerTransactionQuery,
+        ?PaymentUnzerTransactionConditionsTransfer $paymentUnzerTransactionConditionsTransfer
```
there is no reason to pass nullable criteria
please move `if` condition out of this method and call it only when you have conditions to apply

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/UnzerConfig.php:
```diff
@@ -163,6 +168,21 @@ class UnzerConfig extends AbstractBundleConfig
     ];
 
+    /**
+     * @var array<string, string>
```
api tag is missing
same below

kkicheglovspryker replied:
added and sorted public and protected methods

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/UnzerConstants.php:
```diff
@@ -248,4 +263,29 @@ interface UnzerConstants
+    /**
+     * @var string
+     */
+    public const UNZER_MARKETPLACE_BASKET_SHIPMENT_REFERENCE_ID = self::UNZER_BASKET_SHIPMENT_REFERENCE_ID . '_%s';
```
it probably worth to have method for this case
better for extendability

kkicheglovspryker replied:
There is no sense for extending cause it is linked to Unzer API call. I have renamed it for better readability.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/CreateUnzerCredentialsFacadeTest.php:
```diff
@@ -8,8 +8,17 @@
+ * @group Business
+ * @group Facade
+ * @group CreateUnzerCredentialsFacadeTest
+ */
```
please create UnzerFacadeFolder and put all test classes inside
get rid from `facade` naming
e.g. `CreateUnzerCredentialsTest.php`

DmitryLymarenko replied:
still not fixed

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/CreateUnzerCredentialsFacadeTest.php:
```diff
@@ -18,12 +27,13 @@ class CreateUnzerCredentialsFacadeTest extends UnzerFacadeBaseTest
     public function testCreateUnzerCredentials(): void
     {
         //Arrange
```
```suggestion
        // Arrange
```

same for all other annotations

DmitryLymarenko replied:
still not fixed

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/CreateUnzerCredentialsFacadeTest.php:
```diff
@@ -18,12 +27,13 @@ class CreateUnzerCredentialsFacadeTest extends UnzerFacadeBaseTest
     public function testCreateUnzerCredentials(): void
```
is there a negative case?

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/DeleteUnzerCredentialsFacadeTest.php:
```diff
@@ -15,11 +27,14 @@ class DeleteUnzerCredentialsFacadeTest extends UnzerFacadeBaseTest
         //Arrange
-        $parentUnzerCredentialsTransfer = $this->tester->haveMarketplaceUnzerCredentials($this->tester->haveStore());
-        $this->tester->haveMerchantUnzerCredentials($parentUnzerCredentialsTransfer, '');
+        $marketplaceUnzerCredentials = $this->tester->haveMarketplaceUnzerCredentials();
```
```suggestion
        $unzerCredentialsTransfer = $this->tester->haveMarketplaceUnzerCredentials();
```

same everywhere

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/ExecutePostSaveHookFacadeTest.php:
```diff
@@ -7,55 +7,228 @@
+
+        $saveOrderTransfer = $this->tester->haveOrder(
+            [
+                'unitPrice' => 72350,
```
use `OrderTransfer::UNIT_PRICE` instead
same for similar cases across tests

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/GetUnzerCredentialsCollectionFacadeTest.php:
```diff
@@ -18,20 +28,20 @@ class GetUnzerCredentialsCollectionFacadeTest extends UnzerFacadeBaseTest
+        $unzerCredentialsTransfer = $this->tester->haveStandardUnzerCredentials();
 
+        //Act
         $unzerCredentialsCriteriaTransfer = (new UnzerCredentialsCriteriaTransfer())->setUnzerCredentialsConditions(
```
still part of arrange
Act starts at facade call

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/PerformPaymentMethodsImportTest.php:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * @return void
+     */
+    public function testPerformPaymentMethodsImportTest(): void
```
there are several `if` statements at business logic for this facade
all of them must be covered with tests
there is only one happy case covered

DmitryLymarenko replied:
same for other tests

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/_support/Business/UnzerFacadeBaseTest.php:
```diff
@@ -191,14 +199,34 @@ protected function createUtilTextServiceMock()
     protected function getMerchantFacade(): UnzerToMerchantFacadeInterface
     {
-        return new UnzerToMerchantFacadeBridge($this->createMerchantFacadeMock());
+        return new UnzerToMerchantFacadeBridge(new MerchantFacade());
```
could be achieved through localor

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/_support/Business/UnzerFacadeBaseTest.php:
```diff
@@ -191,14 +199,34 @@ protected function createUtilTextServiceMock()
+     */
+    protected function getPaymentFacade(): UnzerToPaymentFacadeInterface
+    {
+        return new UnzerToPaymentFacadeBridge(new PaymentFacade());
```
same

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/_support/Business/UnzerFacadeBaseTest.php:
```diff
@@ -191,14 +199,34 @@ protected function createUtilTextServiceMock()
+    /**
+     * @return \PHPUnit\Framework\MockObject\MockObject|\Spryker\Zed\Payment\Business\PaymentFacade
+     */
+    protected function createPaymentFacadeMock(): PaymentFacade
```
```suggestion
    protected function createPaymentFacadeMock(): PaymentFacadeInterface
```

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/_support/UnzerZedTester.php:
```diff
@@ -188,6 +196,11 @@ class UnzerZedTester extends Actor
+    /**
+     * @var \SprykerEco\Zed\Unzer\Business\UnzerFacadeInterface
+     */
+    protected $facade;
```
several constants at this class are unused
cleanup them

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/codeception.yml:
```diff
@@ -8,31 +8,35 @@ paths:
-        - \SprykerTest\Zed\Oms\Helper\OmsHelper
+    Business:
+        path: Business
+        class_name: UnzerZedTester
```
```suggestion
        class_name: UnzerBusinessTester
```

michbeck replied:
Fixed.

kkicheglovspryker committed:
CC-14716 TcL code review fixes

kkicheglovspryker committed:
CC-14716 TcL code review fixes

Commit:
Fixed tests

Commit:
Fixed UnzerMarketplaceRefundMapper::mapItemTransferToUnzerRefundItemTransfer access modifier

Commit:
Fixed codestyle errors

Commit:
Implemented usage of locator

Commit:
Removed empty tests

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Refund/UnzerRefundProcessor.php:
```diff
@@ -0,0 +1,211 @@
+    /**
+     * @var \SprykerEco\Zed\Unzer\Business\Payment\Processor\Refund\Mapper\UnzerMarketplaceRefundMapperInterface
+     */
+    protected $unzerRefundMapper;
```
not used

michbeck replied:
Fixed.

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/UnzerBusinessFactory.php:
```diff
@@ -490,7 +503,7 @@ public function createUnzerPaymentMapper(): UnzerPaymentMapperInterface
     public function createPaymentProcessorResolver(): UnzerPaymentProcessorResolverInterface
     {
-        $unzerPaymentProcessorsCollection = [
+        return new UnzerPaymentProcessorResolver([
```
please also extract this array to a separate method

michbeck replied:
Fixed.

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/UnzerBusinessFactory.php:
```diff
@@ -569,15 +666,94 @@ public function createUnzerMarketplaceCreditCardChargeProcessor(): UnzerChargePr
+    }
+
+    /**
+     * @return array<\Closure>
```
```suggestion
     * @return array<int, \Closure>
```

michbeck replied:
Fixed.

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/UnzerConfig.php:
```diff
@@ -434,4 +308,168 @@ public function getWebhookEventType(): string
+    /**
+     * @var array<int, string>
+     */
+    protected const UNZER_PAYMENT_STATE_OMS_STATUS_MAP = [
```
constants should be placed above methods

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/CheckIsAuthorizeCanceledOmsConditionTest.php:
```diff
@@ -0,0 +1,110 @@
+     */
+    public function testCheckIsAuthorizeCanceledOmsConditionReturnsTrueWhileOmsStatusIsAuthorizeCanceled(): void
+    {
+        //Arrange
```
```suggestion
        // Arrange
```

same everywhere as I asked before

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/CheckIsAuthorizeCanceledOmsConditionTest.php:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * @var string
+     */
+    public const STATE_MACHINE_PROCESS_NAME = 'UnzerSofort01';
```
```suggestion
    protected const STATE_MACHINE_PROCESS_NAME = 'UnzerSofort01';
```

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/CheckIsAuthorizeCanceledOmsConditionTest.php:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * @var string
+     */
+    public const STATE_MACHINE_PROCESS_NAME = 'UnzerSofort01';
```
```suggestion
    protected const STATE_MACHINE_PROCESS_NAME = 'UnzerSofort01';
```

DmitryLymarenko replied:
same for other tests

michbeck replied:
Fixed.

Commit:
Fixed deprecated call of addFkPaymentUnzerId

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/CheckIsAuthorizeCanceledOmsConditionTest.php:
```diff
@@ -0,0 +1,110 @@
+ * For full license information, please view the LICENSE file that was distributed with this source code.
+ */
+
+namespace SprykerEcoTest\Zed\Unzer\Business;
```
I still don't see folder I requested to add

michbeck replied:
Fixed.

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,200 @@
+                continue;
+            }
+
+            foreach ($orderTransfer->getItems() as $itemTransfer) {
```
please extract inner foreach to a separate method

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/UpdateUnzerCredentialsFacadeTest.php:
```diff
@@ -7,6 +7,18 @@
+ * @group Facade
+ * @group UpdateUnzerCredentialsFacadeTest
+ */
 class UpdateUnzerCredentialsFacadeTest extends UnzerFacadeBaseTest
```
`Facade` postfix still exist
subfloder is still missing
same for other tests

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/UpdateUnzerCredentialsFacadeTest.php:
```diff
@@ -7,6 +7,18 @@
+ * @group Facade
+ * @group UpdateUnzerCredentialsFacadeTest
+ */
 class UpdateUnzerCredentialsFacadeTest extends UnzerFacadeBaseTest
```
`Facade` postfix still exist
subfloder is still missing
same for other tests

michbeck replied:
Fixed.

Commit:
Fixed test group annotations

Commit:
Fixed directory hierarchy

kkicheglovspryker committed:
CC-14716 TcL code review fixes

kkicheglovspryker committed:
Merge branch 'feature/cc-14610/cc-14716-order-commands' of github.com:spryker-eco/unzer into feature/cc-14610/cc-14716-order-commands

Commit:
Adjusted names of test classes

Commit:
Merge remote-tracking branch 'origin/feature/cc-14610/cc-14716-order-commands' into feature/cc-14610/cc-14716-order-commands

kkicheglovspryker committed:
CC-14716 TcL code review fixes

kkicheglovspryker committed:
Merge branch 'feature/cc-14610/cc-14716-order-commands' of github.com:spryker-eco/unzer into feature/cc-14610/cc-14716-order-commands

Commit:
Fixed whitespace

Commit:
Adjusted names of test classes

Commit:
Extracted inner foreach to seperate method

Commit:
Fixed usage of constants within tests

kkicheglovspryker committed:
CC-14716 TcL code review fixes

kkicheglovspryker committed:
CC-14716 TcL code review fixes

Commit:
Fixed codestyle errors

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/Payment/Processor/Charge/UnzerMarketplaceCreditCardChargeProcessor.php:
```diff
@@ -7,57 +7,222 @@
+            if ($itemTransfer->getUnzerParticipantId() === $participantId && $itemTransferFkSalesExpense === $expenseTransferFkSalesExpense) {
+                $expensesAmount += $expenseTransfer->getSumGrossPrice();
+
+                break;
```
return?

michbeck replied:
Fixed.

DmitryLymarenko commented on src/SprykerEco/Zed/Unzer/Business/UnzerBusinessFactory.php:
```diff
@@ -569,15 +674,94 @@ public function createUnzerMarketplaceCreditCardChargeProcessor(): UnzerChargePr
+    /**
+     * @return array<int, \Closure>
+     */
+    public function createUnzerExpenseRefundStrategies(): array
```
```suggestion
    public function getUnzerExpenseRefundStrategies(): array
```

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/UnzerFacade/CreateUnzerCredentialsTest.php:
```diff
@@ -0,0 +1,43 @@
+            ->createUnzerCredentials($unzerCredentialsTransfer);
+
+        // Assert
+        $this->assertTrue($unzerCredentialsResponseTransfer->getIsSuccessful());
```
`$unzerCredentialsResponseTransfer::unzerCredentials` must be asserted as well
the same for similar tests

michbeck replied:
Fixed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/UnzerFacade/CreateUnzerCredentialsTest.php:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * @return void
+     */
+    public function testCreateUnzerCredentials(): void
```
there is a possibility to get non-successful response
we need to cover it as well

michbeck replied:
Fixed.

DmitryLymarenko reviewed (commented):
please also cover cases with exception throwing
I saw multiple places behind the facade
use `$this->expectEcxeption()` for these cases
the same for required transfer properites

kkicheglovspryker committed:
CC-14716 TcL code review fixes

kkicheglovspryker committed:
Merge branch 'feature/cc-14610/cc-14716-order-commands' of github.com:spryker-eco/unzer into feature/cc-14610/cc-14716-order-commands

Commit:
Fixed psalm errors

Commit:
Merge remote-tracking branch 'origin/feature/cc-14610/cc-14716-order-commands' into feature/cc-14610/cc-14716-order-commands

Commit:
Updated CreateUnzerCredentialsTest & DeleteUnzerCredentialsTest

kkicheglovspryker committed:
CC-14716 Added negative unzer api mock

kkicheglovspryker committed:
Merge branch 'feature/cc-14610/cc-14716-order-commands' of github.com:spryker-eco/unzer into feature/cc-14610/cc-14716-order-commands

kkicheglovspryker committed:
CC-14716 Added negative unzer api mock

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/UnzerFacade/CreateUnzerCredentialsTest.php:
```diff
@@ -0,0 +1,297 @@
+
+        // Assert
+        $this->expectException(NullValueException::class);
+        $this->expectExceptionMessage('Property "parentIdUnzerCredentials" of transfer `Generated\Shared\Transfer\UnzerCredentialsTransfer` is null.');
```
this check is not needed
previous is enough

michbeck replied:
Removed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/UnzerFacade/CreateUnzerCredentialsTest.php:
```diff
@@ -0,0 +1,297 @@
+
+        // Assert
+        $this->expectException(NullValueException::class);
+        $this->expectExceptionMessage('Property "unzerKeypair" of transfer `Generated\Shared\Transfer\UnzerCredentialsTransfer` is null.');
```
this check is not needed
previous is enough

michbeck replied:
Removed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/UnzerFacade/CreateUnzerCredentialsTest.php:
```diff
@@ -0,0 +1,297 @@
+
+        // Assert
+        $this->expectException(NullValueException::class);
+        $this->expectExceptionMessage('Property "publicKey" of transfer `Generated\Shared\Transfer\UnzerKeypairTransfer` is null.');
```
this check is not needed
previous is enough

michbeck replied:
Removed.

DmitryLymarenko commented on tests/SprykerEcoTest/Zed/Unzer/Business/UnzerFacade/CreateUnzerCredentialsTest.php:
```diff
@@ -0,0 +1,297 @@
+
+        // Assert
+        $this->expectException(NullValueException::class);
+        $this->expectExceptionMessage('Property "privateKey" of transfer `Generated\Shared\Transfer\UnzerKeypairTransfer` is null.');
```
same everywhere
for transfer property asserting  `$this->expectException(NullValueException::class);` is enough
everything else could be skipped

michbeck replied:
Removed.

Commit:
Removed checks for exception messages

Commit:
Fixed codestyle errors
